cmd/libsnap-confine-private/cgroup-pids-support.c

```diff
@@ -0,0 +1,57 @@
+// For AT_EMPTY_PATH and O_PATH
+#define _GNU_SOURCE
+
+#include "cgroup-pids-support.h"
+
+#include <errno.h>
+#include <fcntl.h>
+#include <limits.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "cleanup-funcs.h"
+#include "string-utils.h"
+#include "utils.h"
+
+static const char *pids_cgroup_dir = "/sys/fs/cgroup/pids";
+
+void sc_cgroup_pids_join(const char *snap_security_tag, pid_t pid) {
+    // Open the pids cgroup directory.
+    int cgroup_fd SC_CLEANUP(sc_cleanup_close) = -1;
+    cgroup_fd = open(pids_cgroup_dir, O_PATH | O_DIRECTORY | O_NOFOLLOW | O_CLOEXEC);
+    if (cgroup_fd < 0) {
+        die("cannot open pids cgroup (%s)", pids_cgroup_dir);
+    }
+    // Create the pid hierarchy for the given snap security tag.
+    if (mkdirat(cgroup_fd, snap_security_tag, 0755) < 0 && errno != EEXIST) {
+        die("cannot create pids cgroup hierarchy for snap security tag %s", snap_security_tag);
+    }
+    // Open the hierarchy directory for the given snap security tag.
+    int hierarchy_fd SC_CLEANUP(sc_cleanup_close) = -1;
+    hierarchy_fd = openat(cgroup_fd, snap_security_tag, O_PATH | O_DIRECTORY | O_NOFOLLOW | O_CLOEXEC);
+    if (hierarchy_fd < 0) {
+        die("cannot open pids cgroup hierarchy for snap security tag %s", snap_security_tag);
+    }
+    // Since we may be running from a setuid but not setgid executable, ensure
+    // that the group and owner of the hierarchy directory is root.root.
+    if (fchownat(hierarchy_fd, "", 0, 0, AT_EMPTY_PATH) < 0) {
```

Comment:
Between the mkdirat() and this fchownat() there is a small window where the owner will not be root:root - is that a cause for concern? If so, should we set permissions to 0700 and only after the chown to the real 0755?

Reply:
There's no cause for concern since there is no g+w bit set. There is nothing inside that is of any sensitivity.